rohmu/delta/snapshot.py

```diff
@@ -74,38 +75,58 @@ def _list_files(self, basepath: Path):
 
         return sorted(result_files)
 
-    def _list_dirs_and_files(self, basepath: Path):
+    def _list_dirs_and_files(self, basepath: Path) -> Tuple[List[Path], List[Path]]:
         files = self._list_files(basepath)
         dirs = {p.parent for p in files}
         return sorted(dirs), files
 
-    def _add_snapshotfile(self, snapshotfile: SnapshotFile):
+    def _add_snapshotfile(self, snapshotfile: SnapshotFile) -> None:
         old_snapshotfile = self.relative_path_to_snapshotfile.get(snapshotfile.relative_path, None)
         if old_snapshotfile:
             self._remove_snapshotfile(old_snapshotfile)
         self.relative_path_to_snapshotfile[snapshotfile.relative_path] = snapshotfile
         if snapshotfile.hexdigest:
             self.hexdigest_to_snapshotfiles.setdefault(snapshotfile.hexdigest, []).append(snapshotfile)
 
-    def _remove_snapshotfile(self, snapshotfile: SnapshotFile):
+    def _remove_snapshotfile(self, snapshotfile: SnapshotFile) -> None:
         assert self.relative_path_to_snapshotfile[snapshotfile.relative_path] == snapshotfile
         del self.relative_path_to_snapshotfile[snapshotfile.relative_path]
         if snapshotfile.hexdigest:
             self.hexdigest_to_snapshotfiles[snapshotfile.hexdigest].remove(snapshotfile)
 
-    def _snapshotfile_from_path(self, relative_path):
+    def _snapshotfile_from_path(self, relative_path: Path, missing_ok: bool) -> SnapshotFile:
         src_path = self.src / relative_path
         st = src_path.stat()
-        return SnapshotFile(relative_path=relative_path, mtime_ns=st.st_mtime_ns, file_size=st.st_size, stored_file_size=0)
+        return SnapshotFile(
+            relative_path=relative_path,
+            mtime_ns=st.st_mtime_ns,
+            file_size=st.st_size,
+            stored_file_size=0,
+            missing_ok=missing_ok,
+        )
 
-    def _gen_snapshot_hashes(self, relative_paths, reuse_old_snapshotfiles):
+    def _gen_snapshot_hashes(
+        self, relative_paths: Sequence[Path], reuse_old_snapshotfiles: bool, required_paths: Optional[Set[Path]] = None
+    ) -> Generator[SnapshotFile, None, None]:
         same = 0
         lost = 0
+
+        if required_paths:
```

Comment:
Are the paths guaranteed to have the same semantics, or required_paths could be absolute while relative_paths could be... relative? AFAIK in pathlib that if p1.absolute() == p2.absolute(), then p1 doesn't necessarily == p2 if they have a different representation. If they could have a different representation we'd need to normalize them somehow before doing those checks.